### ISSUE TITLE
Question about metaaug implementation

### ISSUE THREAD

**Issue by kai-wen-yang** [lionellee9089/metaug#1]
In step 2, when updating the metaaug parameters, why we compute the fast weight for metaaug parameter?
https://github.com/lionellee9089/metaug/blob/98d41bafebfdaa4a4efa018c17f4517366e241a9/train_MetAug.py#L355

In Algorithm 1 of the paper (The line after # compute fast weights # retain computational graph), we compute the fast weight for backbone and head parameter.
![截屏2022-11-12 12 49 52](https://user-images.githubusercontent.com/92202424/201457642-8ac638f4-6631-4a6d-a92a-7a705b7d207b.png)

**Comment by kai-wen-yang** [comment]
Further more, what is step 3 doing in the code?

https://github.com/lionellee9089/metaug/blob/98d41bafebfdaa4a4efa018c17f4517366e241a9/train_MetAug.py#L418

In the paper, it does not have step 3.

**Comment by kai-wen-yang** [comment]
In my understanding, we should compute fast weight for backbone and head parameter, and we compute second-derivative for metaaug parameters.

**Comment by jiangmengli** [comment]
The secondary gradient of augmenter in the code is an attempt of our trick, which itself should be the secondary gradient of the main model (encoder and projection head). We empirically found that end-to-end quadratic gradients are easier to converge, and such a training process leads to the better model performance.

Step 3 is as same as the regular contrastive training step (Step 1) in pseudo code, since we need to ensure the updated augmenter works on the same batches.

**Comment by kai-wen-yang** [comment]
I think it cannot be called a "trick", because the secondary gradient  is an important part of the method, and it seems that the code of this part is very different from the paper.  And this ''trick'' does not make any sense for me, why we first compute fast gradient for augmenter, and then compute secondary gradient for augmenter again in step 2. The paper seems have not mentioned the trick at all. The algorithm in the paper makes sense: it is a meta-learning pipeline, we first compute fast gradient for main model, and then compute secondary gradient for augmenter. I can understand the paper but cannot understand the code, what does the "trick" mean and how it works?

**Comment by kai-wen-yang** [comment]
And you say step 3 is the same as step 1, but step 3 computes secondary gradient while step 1 does not. I cannot understand how step 3 works, why we compute fast gradient for main model, and then compute secondary gradient for main model, what does it mean?

**Comment by jiangmengli** [comment]
Our idea behind the tricks in Step 2 and Step 3 is the ``self-improving''. Refer to MAML and you'll find that there only exists ``one'' model, and the intuition of the mechanism is that the fast weights on sub tasks are the ``trial'' steps, and the second-gradients of the main model based on these fast weights aim to ``improve'' the main model to learn general knowledge (meta knowledge), i.e., the well-designed initial parameters, that can be trained in a few steps to achieve remarkable performance on all sub tasks. The idea of MAML is to improve the main model to learn general knowledge among tasks in a ``self-improving'' manner (since there is no other models). Inspired by this behavior, I use the ``self-improving'' technique to avoid the meta augmenter and the main model falling into local optimum and speed up the training process, which can be done due to the gradients of these networks are computed based on the fast weights of both meta augmenter and the main model. So this gradient can perfectly simulate the real gradient scenario in the next training step thereby leading the training to jointly be speed up and avoid falling into local optimum. 

After knowing the idea behind this behavior, I'll review the processes of Step 2 and Step 3. In Step 2, the gradients of meta augmenter can be split into two parts: 1) the second-gradient of meta augmenter that is computed based on the fast weights of the main model, since the fast weights are computed based on the meta augmenter; 2) this part ought to be the direct gradient of meta augmenter, but due to our trick, we change it into the gradient computed by the ``self-improving'' process, i.e., the second-gradient of meta augmenter itself. The idea is shared in Step 3 either, for the main model, the first part of the loss is as same as Step 1, and the second part of this loss is computed based on the ``self-improving'' technique. The reasons of why we are inclined not to include this technique as a contribution of our paper are: 1) the ``self-improving'' technique is actually proposed by MAML, though MAML didn't call it as a specific name; 2) according to the acceleration technique of such a second-gradient training proposed by MAXL (also not appearing in the paper but existing in the code), i.e., the first-order approximation, the self-improving process can be partly approximated by the direct gradient back-propagation. Therefore, the training process in our code can be approximated by the implementation in our paper. At last, the code and this self-improving technique can slightly improve the model performance but not significant in several datasets, so we don't think this trick is very significant and only provide it in the code for readers to know the existence of our try. I recommend the readers to take a try of this technique in other fields and check whether it can significantly work. Thanks for your comments, and if there is any question, let me know please!

**Comment by kai-wen-yang** [comment]
Yes, the gradients of meta augmenter can be split into two parts: (1) the second-gradient of meta augmenter that is computed based on the fast weights of the main model, and (2) direct gradient of meta augmenter.

But part 1 does not exist in the code: you did not even compute  fast weights of the main model in step 2.

And for part 2, I think the implementation is wrong, it is the direct gradient of augmenter w.r.t the loss, so we did not need to compute second-gradient as the "self-improving" does.

I do not agree that MAML is the same as you, could you please provide the code link you followed? IMO, MAML computes fast weights of main model, and then computes second-gradient of meta parameters w.r.t fast weights of main model.

**Comment by jiangmengli** [comment]
A mistake statement. Actually, Step 2 is the self-improving for meta augmenter. For the self-improving implementation, you can refer to the official implementation, as follows:
![image](https://user-images.githubusercontent.com/15098119/201488738-705e4c5a-92d4-47a2-8b35-61129d6fd22d.png)
The \theta^\prime is the computed parameter based on \theta for the sub-task, and the real updating is performed in line 8, so that there is only one main model, and the corresponding training objective for the main model is to learn task-agnostic (or model-agnostic, since a model for a task) init network parameters.

**Comment by jiangmengli** [comment]
For the part 1), I'm sorry, when I delete the remarks of the code, I delete some specific parts by mistake. The pseudo code in my paper proves that in our original implementation in Step 2 the fast weight of the main model is computed and added into the loss computation. To resolve my mistake, I'll upload both the original version of my code and the version using the self-improving technique soon.

**Comment by kai-wen-yang** [comment]
Thanks for your reply. Now I understand that gradient part 2 (self-improving) seems like to find a better initialization. And I look forward to the update of gradient part 1. I am really interested in how part 1 and part 2 contributes to the performance.
